lib/Analysis/ControlFlowLoopAnalysis.cpp

```diff
@@ -0,0 +1,182 @@
+// expressed as a CFG.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/ControlFlowLoopAnalysis.h"
+#include "mlir/IR/BuiltinOps.h"
+
+using namespace mlir;
+using namespace circt::analysis;
+
+namespace {
+// The BFS callback provides the current block as an argument and returns
+// whether search should halt.
+enum class BFSNextState { Halt, SkipSuccessors, Continue, Custom };
+using BFSCallbackExtended = llvm::function_ref<BFSNextState(
+    Block *, DenseSet<Block *> &, SmallVector<Block *> &)>;
+
+void blockBFS(Block *start, BFSCallbackExtended callback) {
+  DenseSet<Block *> visited;
+  SmallVector<Block *> queue = {start};
+  while (!queue.empty()) {
+    Block *currBlock = queue.front();
+    queue.erase(queue.begin());
+    if (visited.contains(currBlock))
+      continue;
+    visited.insert(currBlock);
+
+    switch (callback(currBlock, visited, queue)) {
+    case BFSNextState::Halt:
+      return;
+    case BFSNextState::Continue: {
+      llvm::copy(currBlock->getSuccessors(), std::back_inserter(queue));
+      break;
+    }
+    case BFSNextState::SkipSuccessors:
+      for (auto *succ : currBlock->getSuccessors())
+        visited.insert(succ);
+      break;
+    case BFSNextState::Custom:
+      break;
+    }
+  }
+}
+
+using BFSCallback = llvm::function_ref<BFSNextState(Block *)>;
+void blockBFS(Block *start, BFSCallback callback) {
+  blockBFS(start, [&](Block *block, DenseSet<Block *> &,
+                      SmallVector<Block *> &) { return callback(block); });
+}
+
+/// Performs a BFS to determine whether there exists a path between 'from' and
+/// 'to'.
+static bool isReachable(Block *from, Block *to) {
+  bool isReachable = false;
+  blockBFS(from, [&](Block *currBlock) {
+    if (currBlock == to) {
+      isReachable = true;
+      return BFSNextState::Halt;
+    }
+    return BFSNextState::Continue;
+  });
+  return isReachable;
+}
+
+} // namespace
+
+/// Helper that checks if entry is a loop header. If it is, it collects
+/// additional information about the loop for further processing.
+LogicalResult ControlFlowLoopAnalysis::collectLoopInfo(Block *entry,
+                                                       LoopInfo &loopInfo) {
+  loopInfo.loopHeader = entry;
+
+  for (auto *backedge : entry->getPredecessors()) {
+    bool dominates = domInfo.dominates(entry, backedge);
+    bool formsLoop = isReachable(entry, backedge);
+    if (formsLoop) {
+      if (dominates)
+        loopInfo.loopLatches.insert(backedge);
+      else
+        return entry->getParentOp()->emitError()
+               << "Non-canonical loop structures detected; a potential "
+                  "loop header has backedges not dominated by the loop "
+                  "header. This indicates that the loop has multiple entry "
+                  "points. Handshake lowering does not yet support this "
+                  "form of control flow, exiting.";
```

Comment:
I assume this is still an error in that we expect to gather canonical loop structures - but the last sentence should probably be removed 😛.

Reply:
Good catch. Yes, this is indeed still an error, but the reference to the handshake lowering should indeed be dropped.